include/seqan3/alignment/matrix/alignment_score_matrix.hpp

```diff
@@ -0,0 +1,156 @@
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//     * Neither the name of Knut Reinert or the FU Berlin nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+/*!\file
+ * \author Marcel Ehrhardt <marcel.ehrhardt AT fu-berlin.de>
+ * \brief Contains the declaration of seqan3::alignment_score_matrix.
+ */
+
+#pragma once
+
+#include <vector>
+#include <seqan3/core/platform.hpp>
+
+namespace seqan3
+{
+
+//!\brief The declaration of alignment_score_matrix. Each definition of this
+//!       declaration must satisfy seqan3::alignment_matrix_concept.
+//!\ingroup alignment_matrix
+template <typename ...>
+struct alignment_score_matrix;
+
+/*!\brief A score matrix represented in a one-dimensional std::vector
+ * \ingroup alignment_matrix
+ * \implements seqan3::alignment_matrix_concept
+ * \tparam score_t    \copydoc seqan3::alignment_score_matrix<std::vector<score_t>,sequence_t>::entry_type
+ * \tparam sequence_t \copydoc seqan3::alignment_matrix_concept::sequence_type
+ *
+ *
+ * This data structure stores the matrix in a flat way using the
+ * std::vector<#entry_type> data structure where each row is stored
+ * continuously.
+ *
+ * ## Example
+ *
+ * \snippet test/snippet/alignment/matrix/alignment_score_matrix.cpp code
+ *
+ * ### Output
+ * \include test/snippet/alignment/matrix/alignment_score_matrix.out
+ */
+template <typename score_t, typename sequence_t>
+//!\cond
+    requires std::is_integral_v<score_t>
+//!\endcond
+struct alignment_score_matrix<std::vector<score_t>, sequence_t>
```

Comment:
I don't see, why we need specialisations of it. 
Further, I think the would like to specify the vector type as a template template argument, which defaults to std::vector. 
But, we want to be able to specify the allocator for the matrix (like mem_aligned allocator for SIMD stuff).
It could be possible that we want to have a stack matrix using std::array as container?
In this case we should think about having a traits class rather than the current type specification. 
These are just thoughts. We should discuss this later.